core/src/main/scala/org/apache/spark/sql/execution/columnar/impl/RemoteEntriesIterator.scala

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright (c) 2018 SnappyData, Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you
+ * may not use this file except in compliance with the License. You
+ * may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied. See the License for the specific language governing
+ * permissions and limitations under the License. See accompanying
+ * LICENSE file.
+ */
+package org.apache.spark.sql.execution.columnar.impl
+
+import scala.collection.JavaConverters._
+
+import com.gemstone.gemfire.internal.cache.{NonLocalRegionEntry, PartitionedRegion, RegionEntry, TXState, Token}
+import com.pivotal.gemfirexd.internal.engine.distributed.message.GetAllExecutorMessage
+import com.pivotal.gemfirexd.internal.engine.sql.execute.GemFireResultSet
+import io.snappydata.collection.LongObjectHashMap
+
+import org.apache.spark.sql.execution.columnar.impl.ColumnFormatEntry.{DELETE_MASK_COL_INDEX, STATROW_COL_INDEX}
+
+/**
+ * A [[ClusteredColumnIterator]] that fetches entries from a remote bucket.
+ */
+final class RemoteEntriesIterator(bucketId: Int, projection: Array[Int],
+    pr: PartitionedRegion, txState: TXState) extends ClusteredColumnIterator {
+
+  private val statsRows = {
+    val statsKeys = pr.getBucketKeys(bucketId).iterator().asScala.filter {
+      case k: ColumnFormatKey => k.columnIndex == STATROW_COL_INDEX
+      case _ => false
+    }
+    // get the full entries for all stats rows in one getAll
+    val msg = new GetAllExecutorMessage(pr, statsKeys.asInstanceOf[Iterator[AnyRef]].toArray,
+      null, null, null, null, null, null, txState, null, false, false)
+    statsKeys.zip(GemFireResultSet.callGetAllExecutorMessage(msg).asScala.toIterator)
+  }
+
+  /**
+   * Full projection including all of delta and meta-data columns (except base stats entry)
+   */
+  private val fullProjection = {
```

Comment:
So the projection list is passed so that only the required columns are fetched. Now the delta buffer and the bit masks also needs to be brought I understand. But can you please explain this maths or add a comment in the code itself that explains what all is being desired to be brought.

(projection.length * (ColumnDelta.USED_MAX_DEPTH + 1) - DELETE_MASK_COL_INDEX + STATROW_COL_INDEX)

Reply:
The math comes directly from the loops below that fill it. Will add a comment nevertheless.